chore(server): don't insert embeddings if the model has changed

uhthomas opened the pull request:
## Description

We're moving away from the heuristic of waiting for queues to complete. The job which inserts embeddings can simply check if the model has changed before inserting, rather than attempting to lock the queue.
Fixes # (issue)

## How Has This Been Tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

N/A

<!-- API endpoint changes (if relevant)
## API Changes
The `/api/something` endpoint is now `/api/something-else`
-->

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if applicable
- [x] I have no unrelated changes in the PR.
- [x] I have confirmed that any new dependencies are strictly necessary.
- [x] I have written tests for new code (if applicable)
- [x] I have followed naming conventions/patterns in the surrounding code
- [x] All code in `src/services/` uses repositories implementations for database calls, filesystem operations, etc.
- [x] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services/`)

mertalev commented on server/src/services/smart-info.service.ts:
```diff
@@ -126,6 +119,13 @@ export class SmartInfoService extends BaseService {
       await this.databaseRepository.wait(DatabaseLock.CLIPDimSize);
     }
 
+    const newConfig = await this.getConfig({ withCache: false });
+    if (machineLearning.clip.modelName !== newConfig.machineLearning.clip.modelName) {
+      // Skip the job if the the model has changed since the embedding was
+      // generated.
+      return JobStatus.SKIPPED;
+    }
```
I wonder if the check can be done with events instead. It seems a bit wasteful to check this for every job when changing the model is such a rare occasion.

uhthomas replied:
Maybe - how would that look?

uhthomas replied:
Updated.

mertalev reviewed (commented):
Can you test the behavior when switching a model a few times while there are active smart search jobs? There shouldn't be any DB errors in the logs. Also if using a config file and the server starts up.

uhthomas committed:
chore(server): don't insert embeddings if the model has changed

We're moving away from the heuristic of waiting for queues to complete. The job
which inserts embeddings can simply check if the model has changed before
inserting, rather than attempting to lock the queue.

mertalev commented:
LGTM but needs testing before merge

uhthomas commented on server/src/services/smart-info.service.ts:
```diff
@@ -126,6 +119,13 @@ export class SmartInfoService extends BaseService {
       await this.databaseRepository.wait(DatabaseLock.CLIPDimSize);
     }
 
+    const newConfig = await this.getConfig({ withCache: true });
```
I wonder if it's worth doing this check in a transaction?

mertalev replied:
Why?

uhthomas replied:
I suppose that even though it's unlikely, the config could change, and something else then do the truncation and stuff in between this check and actually inserting.

mertalev committed:
more robust dim size update

mertalev committed:
use check constraint

mertalev committed:
index command cleanup

mertalev committed:
add create statement

mertalev committed:
update medium test, create appropriate extension

mertalev committed:
new line

mertalev committed:
set dimension size when running on all assets

mertalev committed:
why does it want braces smh

mertalev committed:
take 2

mertalev reviewed (approved):
Tested locally by switching models repeatedly while many smart search jobs are queued and actively running. There are 1-2 rejected embeddings with a job concurrency of 16 and the transition otherwise has no errors. The index and dimension size are updated as expected and searching immediately after works as well. Some of the changes here are also align with the upcoming VectorChord changes. I did kind of hijack this PR, so would be good to get another pair of eyes.

uhthomas commented on server/src/migrations/1700713994428-AddCLIPEmbeddingIndex.ts:
@mertalev existing migrations can't be changed, this will require a new migration.

mertalev replied:
The behavior is the same, just consolidating the indexing command to one place.

uhthomas replied:
Gotcha, thank you for clarifying

jrasm91 reviewed (approved):
LGTM
